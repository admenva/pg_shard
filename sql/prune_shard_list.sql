-- ===================================================================
-- create test functions
-- ===================================================================

CREATE FUNCTION prune_using_no_values(regclass)
	RETURNS text[]
	AS 'pg_shard'
	LANGUAGE C STRICT;

CREATE FUNCTION prune_using_single_value(regclass, text)
	RETURNS text[]
	AS 'pg_shard'
	LANGUAGE C;

CREATE FUNCTION prune_using_either_value(regclass, text, text)
	RETURNS text[]
	AS 'pg_shard'
	LANGUAGE C STRICT;

CREATE FUNCTION prune_using_both_values(regclass, text, text)
	RETURNS text[]
	AS 'pg_shard'
	LANGUAGE C STRICT;

CREATE FUNCTION debug_equality_expression(regclass)
	RETURNS cstring
	AS 'pg_shard'
	LANGUAGE C STRICT;

-- ===================================================================
-- test shard pruning functionality
-- ===================================================================

CREATE TABLE pruning ( species text, last_pruned date, plant_id integer );

INSERT INTO pgs_distribution_metadata.partition (relation_id, partition_method, key)
VALUES
	('pruning'::regclass, 'h', 'species');

INSERT INTO pgs_distribution_metadata.shard
	(id, relation_id, storage, min_value, max_value)
VALUES
	(10, 'pruning'::regclass, 't', '-2147483648', '-1073741826'),
	(11, 'pruning'::regclass, 't', '-1073741825', '-3'),
	(12, 'pruning'::regclass, 't', '-2', '1073741820'),
	(13, 'pruning'::regclass, 't', '1073741821', '2147483647');

SELECT prune_using_no_values('pruning');

SELECT prune_using_single_value('pruning', 'tomato');

SELECT prune_using_single_value('pruning', NULL);

SELECT prune_using_either_value('pruning', 'tomato', 'petunia');

SELECT prune_using_both_values('pruning', 'tomato', 'petunia');

SELECT prune_using_both_values('pruning', 'tomato', 'rose');

SELECT debug_equality_expression('pruning');
