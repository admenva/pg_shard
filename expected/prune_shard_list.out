-- ===================================================================
-- create test functions
-- ===================================================================
CREATE FUNCTION prune_using_no_values(regclass)
	RETURNS text[]
	AS 'pg_shard'
	LANGUAGE C STRICT;
CREATE FUNCTION prune_using_single_value(regclass, text)
	RETURNS text[]
	AS 'pg_shard'
	LANGUAGE C;
CREATE FUNCTION prune_using_either_value(regclass, text, text)
	RETURNS text[]
	AS 'pg_shard'
	LANGUAGE C STRICT;
CREATE FUNCTION prune_using_both_values(regclass, text, text)
	RETURNS text[]
	AS 'pg_shard'
	LANGUAGE C STRICT;
CREATE FUNCTION debug_equality_expression(regclass)
	RETURNS cstring
	AS 'pg_shard'
	LANGUAGE C STRICT;
-- ===================================================================
-- test shard pruning functionality
-- ===================================================================
CREATE TABLE pruning ( species text, last_pruned date, plant_id integer );
INSERT INTO pgs_distribution_metadata.partition (relation_id, partition_method, key)
VALUES
	('pruning'::regclass, 'h', 'species');
INSERT INTO pgs_distribution_metadata.shard
	(id, relation_id, storage, min_value, max_value)
VALUES
	(10, 'pruning'::regclass, 't', '-2147483648', '-1073741826'),
	(11, 'pruning'::regclass, 't', '-1073741825', '-3'),
	(12, 'pruning'::regclass, 't', '-2', '1073741820'),
	(13, 'pruning'::regclass, 't', '1073741821', '2147483647');
SELECT prune_using_no_values('pruning');
 prune_using_no_values 
-----------------------
 {13,12,11,10}
(1 row)

SELECT prune_using_single_value('pruning', 'tomato');
 prune_using_single_value 
--------------------------
 {12}
(1 row)

SELECT prune_using_single_value('pruning', NULL);
 prune_using_single_value 
--------------------------
 {12}
(1 row)

SELECT prune_using_either_value('pruning', 'tomato', 'petunia');
 prune_using_either_value 
--------------------------
 {12,11}
(1 row)

SELECT prune_using_both_values('pruning', 'tomato', 'petunia');
 prune_using_both_values 
-------------------------
 {}
(1 row)

SELECT prune_using_both_values('pruning', 'tomato', 'rose');
 prune_using_both_values 
-------------------------
 {12}
(1 row)

SELECT debug_equality_expression('pruning');
                                                                                                                                                                           debug_equality_expression                                                                                                                                                                            
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 {OPEXPR :opno 98 :opfuncid 67 :opresulttype 16 :opretset false :opcollid 0 :inputcollid 100 :args ({VAR :varno 1 :varattno 1 :vartype 25 :vartypmod -1 :varcollid 100 :varlevelsup 0 :varnoold 1 :varoattno 1 :location -1} {CONST :consttype 25 :consttypmod -1 :constcollid 100 :constlen -1 :constbyval false :constisnull true :location -1 :constvalue <>}) :location -1}
(1 row)

